### PR TITLE
support segments synthesis

### PR DESCRIPTION
**Pull request by olasaadi99**
support segments synthesis

### CODE REVIEW

**Comment by olasaadi99** [comment]
> I created a new example with subnet grouping (folder is called `subnet-grouping`) and ran it as follows:
> 
> ```
> ../../vpc-network-config-analyzer/bin/vpcanalyzer report subnets -c config_object.json -o synthesis -f subnets_synthesis_g.json -g
> ```
> 
> This shows two problems:
> 
> 1. The segments are not added to the output fiIe
segments are added to the out file.. at the end of the file..
will fix other problems

**Comment by kyorav** [comment]
> segments are added to the out file.. at the end of the file..

They were not added to the out file when I ran it. Should I pull a fresh version of the branch and try again?

**Comment by olasaadi99** [comment]
> > segments are added to the out file.. at the end of the file..
> 
> They were not added to the out file when I ran it. Should I pull a fresh version of the branch and try again?

I think you should

**Comment by olasaadi99** [comment]
> If I run without `-g` (grouping) then I get `subgrouping/` as a prefix for all subnets:
> 
> ```
> "dst": {
>                 "name": "subgrouping/sub2",
>                 "type": "subnet"
>             }
> ```

it is the name of the vpc... as expected

**Comment by kyorav** [comment]
> it is the name of the vpc... as expected
of course, my bad

The output with grouping looks great. Without grouping it is too long to check in depth, but I can't see any glaring issues.

File: pkg/vpcmodel/synthesisOutput.go [hunk 1]
```diff
@@ -41,35 +45,47 @@ func (j *SynthesisOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
 	return &SingleAnalysisOutput{Output: outStr, VPC1Name: c1.VPC.Name(), VPC2Name: v2Name, format: Synthesis, jsonStruct: all}, err
 }
 
-func handleNameAndType(resource EndpointElem, externals spec.SpecExternals) (
+func handleNameAndType(resource EndpointElem, externals spec.SpecExternals,
+	segments spec.SpecSegments, grouping bool) (
 	resourceName string,
 	resourceType spec.ResourceType) {
```
Comment (zivnevo) [inline pkg/vpcmodel/synthesisOutput.go:51]:
```suggestion
func handleNameAndType(resource EndpointElem, externals spec.SpecExternals, segments spec.SpecSegments, grouping bool) (
	resourceName string, resourceType spec.ResourceType) {
```
too many lines

File: pkg/vpcmodel/synthesisOutput.go [hunk 2]
```diff
@@ -41,35 +45,47 @@ func (j *SynthesisOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
...
 	if resource.IsExternal() {
 		if structObj, ok := resource.(*groupedExternalNodes); ok {
 			// should be always true if src is external"
 			// later in aggregate we change the name with other vpc configs
 			externals[resourceName] = structObj.CidrOrAddress()
 		}
 	}
-	resourceType = resource.SynthesisKind()
+	if grouping && !resource.IsExternal() {
+		if structObj, ok := resource.(*groupedEndpointsElems); ok {
```
Comment (zivnevo) [inline pkg/vpcmodel/synthesisOutput.go:62]:
```suggestion
		if groupObj, ok := resource.(*groupedEndpointsElems); ok {
```
`structObj` is a bit misleading. The type is a slice of structs, not a struct

File: pkg/vpcmodel/synthesisOutput.go [hunk 3]
```diff
@@ -41,35 +45,47 @@ func (j *SynthesisOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
...
+			// should be always true if src is internal"
+			// later in aggregate we change the name with other vpc configs
+			if structObj.LenOfGroupedEndpoints() > 1 {
```
Comment (zivnevo) [inline pkg/vpcmodel/synthesisOutput.go:65]:
```suggestion
			if len(*structObj) > 1 {
```
No need for a dedicated function

File: pkg/vpcmodel/grouping.go [hunk 4]
```diff
@@ -221,8 +221,21 @@ func (g *groupedEndpointsElems) SynthesisResourceName() string {
 	return g.Name()
 }
 
+func (g *groupedEndpointsElems) GroupedEndpointsElemsNamesList() []string {
```
Comment (zivnevo) [inline pkg/vpcmodel/grouping.go:224]:
```suggestion
func (g *groupedEndpointsElems) AsNamesList() []string {
```
Struct type should not be part of a its method name

File: pkg/vpcmodel/output.go [hunk 5]
```diff
@@ -303,38 +304,48 @@ func (of *serialOutputFormatter) AggregateVPCsOutput(outputList []*SingleAnalysi
...
 		// in synthesis format we need to follow json spec schema
-		res, err = writeJSON(spec.Spec{RequiredConnections: connLines, Externals: externals}, outFile)
+		res, err = writeJSON(spec.Spec{Externals: externals, Segments: segments, RequiredConnections: connLines}, outFile)
 	}
 	return res, err
 }
 
-func getNewExternalName(externalName string, externalsMap map[string]string) string {
-	if val, ok := externalsMap[externalName]; ok {
+func getNewExternalSegmentName(externalSegmentName, prefixString string, namesMap map[string]string) string {
```
Comment (zivnevo) [inline pkg/vpcmodel/output.go:327]: Move this function to `synthesisOutput.go`

File: pkg/vpcmodel/output.go [hunk 6]
```diff
@@ -303,38 +304,48 @@ func (of *serialOutputFormatter) AggregateVPCsOutput(outputList []*SingleAnalysi
...
-	name := fmt.Sprintf("%s%d", externalString, len(externalsMap))
-	externalsMap[externalName] = name
+	name := fmt.Sprintf("%s%d", prefixString, len(namesMap))
+	namesMap[externalSegmentName] = name
 	return name
 }
 
-func renameExternals(requiredConnections []spec.SpecRequiredConnectionsElem,
-	externalsMap map[string]string) []spec.SpecRequiredConnectionsElem {
+func renameExternalsSegments(requiredConnections []spec.SpecRequiredConnectionsElem,
```
Comment (zivnevo) [inline pkg/vpcmodel/output.go:336]: Move to `synthesisOutput.go`

Comment (zivnevo) [inline pkg/vpcmodel/output.go:336]:
```suggestion
func renameExternalsAndSegments(requiredConnections []spec.SpecRequiredConnectionsElem,
```

File: pkg/vpcmodel/output.go (hunk 5 above)
Comment (zivnevo) [inline pkg/vpcmodel/output.go:327]:
```suggestion
func getNewExternalOrSegmentName(externalSegmentName, prefixString string, namesMap map[string]string) string {
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
Looks good.
A few small things

**Review by zivnevo — APPROVED** [review]

**Pull request merged**